### ISSUE TITLE
Sparklyr 1.4 Does Not Propagete spark.kubernetes.driver.pod.name to K8s cluster Properly

### ISSUE THREAD

**Issue by mn-mikke** [sparklyr/sparklyr#2832]
Hi,
After an upgrade from Sparklyr 1.3.1 to 1.4 I have been struggling to deploy application on K8s.

```
> library(sparklyr)
> library(rsparkling)
> options(sparklyr.log.console = TRUE)
> sparkHome <-"/Users/marek/software/spark/spark-2.4.7-bin-hadoop2.7"
> config["spark.home"] <- sparkHome
> master = "k8s://https://09D607EDAFD3BB58CB2937855E802F55.yl4.us-west-2.eks.amazonaws.com"
> image = "h2oai/sparkling-water-r:3.32.0.2-1-2.4"
> extraOptionsParsed <- list()
> extraOptionsParsed["spark.kubernetes.file.upload.path"] <-"file:///tmp"
> config <- spark_config_kubernetes(master = master, image = image, account = "spark", driver = "driver-r", version = "2.4.7", executors = 2, conf = extraOptionsParsed, ports = c(8880, 8881, 4040, 54321))
> sc <- spark_connect(config = config, spark_home=sparkHome)
Error from server (NotFound): pods "driver-r" not found
Error in spark_connect_gateway(gatewayAddress, gatewayPort, sessionId,  : 
  Gateway in localhost:8880 did not respond.


Try running `options(sparklyr.log.console = TRUE)` followed by `sc <- spark_connect(...)` for more debugging info.
```

A pod with a different name is cratead instead and fails:
```
> kubectl get pod                           
NAME                            READY   STATUS   RESTARTS   AGE
sparklyr-1606318139222-driver   0/1     Error    0          6m5s
```

Logs from the pod:
```
> kubectl logs sparklyr-1606318139222-driver                             
++ id -u
+ myuid=0
++ id -g
+ mygid=0
+ set +e
++ getent passwd 0
+ uidentry=root:x:0:0:root:/root:/bin/bash
+ set -e
+ '[' -z root:x:0:0:root:/root:/bin/bash ']'
+ SPARK_K8S_CMD=driver
+ case "$SPARK_K8S_CMD" in
+ shift 1
+ SPARK_CLASSPATH=':/opt/spark/jars/*'
+ env
+ sed 's/[^=]*=\(.*\)/\1/g'
+ sort -t_ -k4 -n
+ grep SPARK_JAVA_OPT_
+ readarray -t SPARK_EXECUTOR_JAVA_OPTS
+ '[' -n '' ']'
+ '[' -n '' ']'
+ PYSPARK_ARGS=
+ '[' -n '' ']'
+ R_ARGS=
+ '[' -n '' ']'
+ '[' '' == 2 ']'
+ '[' '' == 3 ']'
+ case "$SPARK_K8S_CMD" in
+ CMD=("$SPARK_HOME/bin/spark-submit" --conf "spark.driver.bindAddress=$SPARK_DRIVER_BIND_ADDRESS" --deploy-mode client "$@")
+ exec /usr/bin/tini -s -- /opt/spark/bin/spark-submit --conf spark.driver.bindAddress=192.168.3.223 --deploy-mode client --properties-file /opt/spark/conf/spark.properties --class sparklyr.Shell spark-internal spark.kubernetes.authenticate.driver.serviceAccountName=spark spark.executor.instances=2 spark.kubernetes.file.upload.path=file:///tmp --jars /Users/marek/Library/R/3.5/library/rsparkling/java/sparkling_water_assembly.jar local:///opt/sparklyr/sparklyr-2.4-2.11.jar 8880 97407 --remote
20/11/25 15:29:05 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
20/11/25 15:29:06 WARN DependencyUtils: Local jar /Users/marek/testing-infra/spark.kubernetes.driver.pod.name=driver-r does not exist, skipping.
20/11/25 15:29:06 WARN SparkSubmit$$anon$2: Failed to load sparklyr.Shell.
java.lang.ClassNotFoundException: sparklyr.Shell
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.apache.spark.util.Utils$.classForName(Utils.scala:238)
	at org.apache.spark.deploy.SparkSubmit.org$apache$spark$deploy$SparkSubmit$$runMain(SparkSubmit.scala:806)
	at org.apache.spark.deploy.SparkSubmit.doRunMain$1(SparkSubmit.scala:161)
	at org.apache.spark.deploy.SparkSubmit.submit(SparkSubmit.scala:184)
	at org.apache.spark.deploy.SparkSubmit.doSubmit(SparkSubmit.scala:86)
	at org.apache.spark.deploy.SparkSubmit$$anon$2.doSubmit(SparkSubmit.scala:920)
	at org.apache.spark.deploy.SparkSubmit$.main(SparkSubmit.scala:929)
	at org.apache.spark.deploy.SparkSubmit.main(SparkSubmit.scala)
20/11/25 15:29:06 INFO ShutdownHookManager: Shutdown hook called
20/11/25 15:29:06 INFO ShutdownHookManager: Deleting directory /tmp/spark-8f28360d-5f6d-471a-996a-38d65f13f1db
```

This line looks suspicius:
```
20/11/25 15:29:06 WARN DependencyUtils: Local jar /Users/marek/testing-infra/spark.kubernetes.driver.pod.name=driver-r does not exist, skipping.
```

(`Users/marek/testing-infra/` is a working directory)

It looks like an error in building of the submit command. If I downgrade Sparklyr to 1.3.1, there is no problem. Please let me know if I need to specify extra parameters in Sparklyr 1.4. Thanks in advance!

**Comment by yitao-li** [comment]
@mn-mikke I think there is a bigger problem: `sparklyr` expects to find a Scala class named `Sparklyr.Shell` from this jar file: local:///opt/sparklyr/sparklyr-2.4-2.11.jar but it's unable to do that. That means `sparklyr` won't be able to submit the Spark app it needs. This may be related to the `Local jar /Users/marek/testing-infra/spark.kubernetes.driver.pod.name=driver-r does not exist` issue you mentioned.

I really don't think anything has changed between sparklyr 1.3.1 and sparklyr 1.4 in term of working with k8s, but I'll need to double-check that is the case.

**Comment by yitao-li** [comment]
@mn-mikke I think https://github.com/sparklyr/sparklyr/pull/2834 should fix it.

Reason to think that should be the case:

after that change and re-installing `sparklyr` from source, the following config

```
library(sparklyr)

config <- list()
sparkHome <-"~/spark/spark-2.4.7-bin-hadoop2.7"
config[["spark.home"]] <- sparkHome
master <- "k8s://https://abcd.efgh.com"
image <- "h2oai/sparkling-water-r:3.32.0.2-1-2.4"
extraOptionsParsed <- list(spark.kubernetes.file.upload.path = "file:///tmp")
config <- spark_config_kubernetes(master = master, image = image, account = "spark", driver = "driver-r", version = "2.4.7", executors = 2, conf = extraOptionsParsed, ports = c(8880, 8881, 4040, 54321))
```

produces a sensible list of spark-submit shell args, i.e.,:

```
print(sparklyr:::spark_config_shell_args(config = config, master = master))
```

outputs

```
 [1] "--master"
 [2] "k8s://https://abcd.efgh.com"
 [3] "--deploy-mode"
 [4] "cluster"
 [5] "--name"
 [6] "sparklyr"
 [7] "--class"
 [8] "sparklyr.Shell"
 [9] "--conf"
[10] "spark.kubernetes.container.image=h2oai/sparkling-water-r:3.32.0.2-1-2.4"
[11] "--conf"
[12] "spark.kubernetes.driver.pod.name=driver-r"
[13] "--conf"
[14] "spark.kubernetes.authenticate.driver.serviceAccountName=spark"
[15] "--conf"
[16] "spark.executor.instances=2"
[17] "--conf"
[18] "spark.kubernetes.file.upload.path=file:///tmp"
```
(we needed multiple `--conf` flags)

**Comment by yitao-li** [comment]
Also assuming the PR above fixes everything, then it will be shipped with sparklyr 1.5.1 shortly.

**Comment by mn-mikke** [comment]
Thanks a lot @yitao-li!
